Add a ClosedArc path segment type

heycam opened the pull request:
#### 2caf219a57a77bfec2cdbbb8c987f186356cfc02
<pre>
Add a ClosedArc path segment type
<a href="https://bugs.webkit.org/show_bug.cgi?id=268236">https://bugs.webkit.org/show_bug.cgi?id=268236</a>
<a href="https://rdar.apple.com/120928268">rdar://120928268</a>

Reviewed by Simon Fraser.

Store paths consisting of a arc() / closePath() pair as a single path
segment, and add a single display list item to represent filling or
stroking such paths.

* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::asSingleArc const):
(WebCore::Path::closeSubpath):
(WebCore::Path::singleClosedArc const):
(WebCore::Path::isClosed const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::PathImpl::singleClosedArc const):
* Source/WebCore/platform/graphics/PathSegment.h:
(WebCore::PathSegment::closesSubpath const):
(WebCore::PathSegment::isCloseSubPath const): Deleted.
* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::operator&lt;&lt;):
(WebCore::PathClosedArc::calculateEndPoint const):
(WebCore::PathClosedArc::tryGetEndPointWithoutContext const):
(WebCore::PathClosedArc::extendFastBoundingRect const):
(WebCore::PathClosedArc::extendBoundingRect const):
* Source/WebCore/platform/graphics/PathSegmentData.h:
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::add):
(WebCore::PathStream::singleClosedArc const):
(WebCore::PathStream::isClosed const):
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::PathCG::add):
(WebCore::addToCGContextPath):
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::FillClosedArc::apply const):
(WebCore::DisplayList::FillClosedArc::dump const):
(WebCore::DisplayList::StrokeClosedArc::apply const):
(WebCore::DisplayList::StrokeClosedArc::dump const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::FillClosedArc::FillClosedArc):
(WebCore::DisplayList::FillClosedArc::closedArc const):
(WebCore::DisplayList::FillClosedArc::path const):
(WebCore::DisplayList::StrokeClosedArc::StrokeClosedArc):
(WebCore::DisplayList::StrokeClosedArc::closedArc const):
(WebCore::DisplayList::StrokeClosedArc::path const):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::fillPath):
(WebCore::DisplayList::Recorder::strokePath):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordFillClosedArc):
(WebCore::DisplayList::RecorderImpl::recordStrokeClosedArc):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebKit/GPUProcess/graphics/PathSegment.serialization.in:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::fillClosedArc):
(WebKit::RemoteDisplayListRecorder::strokeClosedArc):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordFillClosedArc):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeClosedArc):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: <a href="https://commits.webkit.org/273683@main">https://commits.webkit.org/273683@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9258f3977be0a3fef22164a142a54b71fed188a7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36108 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15054 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/38308 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38828 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32463 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37330 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17494 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/12083 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31160 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36663 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12720 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32039 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11146 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/36258 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11171 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32208 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40075 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32785 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32578 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37108 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11360 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9252 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35193 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13086 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8238 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11840 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12198 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/PathSegmentData.cpp:
```diff
@@ -427,6 +427,37 @@ WTF::TextStream& operator<<(WTF::TextStream& ts, const PathArc& data)
+    auto maxXMaxYCorner = center + FloatSize { radius, radius };
+    boundingRect.extend(minXMinYCorner);
+    boundingRect.extend(maxXMaxYCorner);
+}
```
Am I confused here? How can this be correct? I think this is extending the `boundingRect` by the area of the circle which contains the arc. I think we can have this approximation implemented in `extendFastBoundingRect()`. But here I think we need to have more accurate calculation.

heycam replied:
You are right. I was confusing the test content I was looking at (which draws an arc that is a full circle) and the general case.

shallawa commented on Source/WebCore/platform/graphics/PathSegmentData.h:
```diff
@@ -165,6 +165,28 @@ struct PathArc {
 
 WEBCORE_EXPORT WTF::TextStream& operator<<(WTF::TextStream&, const PathArc&);
 
+struct PathClosedArc {
```
Can't this structure contain a single member of type `PathArc' instead of repeating the same data members here?

heycam committed:
Add a ClosedArc path segment type
https://bugs.webkit.org/show_bug.cgi?id=268236
rdar://120928268

Reviewed by Simon Fraser.

Store paths consisting of a arc() / closePath() pair as a single path
segment, and add a single display list item to represent filling or
stroking such paths.

* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::asSingleArc const):
(WebCore::Path::closeSubpath):
(WebCore::Path::singleClosedArc const):
(WebCore::Path::isClosed const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::PathImpl::singleClosedArc const):
* Source/WebCore/platform/graphics/PathSegment.h:
(WebCore::PathSegment::closesSubpath const):
(WebCore::PathSegment::isCloseSubPath const): Deleted.
* Source/WebCore/platform/graphics/PathSegmentData.cpp:
(WebCore::operator<<):
(WebCore::PathClosedArc::calculateEndPoint const):
(WebCore::PathClosedArc::tryGetEndPointWithoutContext const):
(WebCore::PathClosedArc::extendFastBoundingRect const):
(WebCore::PathClosedArc::extendBoundingRect const):
* Source/WebCore/platform/graphics/PathSegmentData.h:
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::add):
(WebCore::PathStream::singleClosedArc const):
(WebCore::PathStream::isClosed const):
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::PathCG::add):
(WebCore::addToCGContextPath):
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::FillClosedArc::apply const):
(WebCore::DisplayList::FillClosedArc::dump const):
(WebCore::DisplayList::StrokeClosedArc::apply const):
(WebCore::DisplayList::StrokeClosedArc::dump const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
(WebCore::DisplayList::FillClosedArc::FillClosedArc):
(WebCore::DisplayList::FillClosedArc::closedArc const):
(WebCore::DisplayList::FillClosedArc::path const):
(WebCore::DisplayList::StrokeClosedArc::StrokeClosedArc):
(WebCore::DisplayList::StrokeClosedArc::closedArc const):
(WebCore::DisplayList::StrokeClosedArc::path const):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::fillPath):
(WebCore::DisplayList::Recorder::strokePath):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::recordFillClosedArc):
(WebCore::DisplayList::RecorderImpl::recordStrokeClosedArc):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebKit/GPUProcess/graphics/PathSegment.serialization.in:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::fillClosedArc):
(WebKit::RemoteDisplayListRecorder::strokeClosedArc):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Shared/DisplayListArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::recordFillClosedArc):
(WebKit::RemoteDisplayListRecorderProxy::recordStrokeClosedArc):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:

Canonical link: https://commits.webkit.org/273683@main
